### PR TITLE
Initial draft for C# (.NET) quickstart

### PR DESCRIPTION
**Pull request by 2percentsilk**
### Why:
To help customers be successful with Codespaces, we're creating a set of language quickstarts.

**Closes [issue link]**

### What's being changed:

Adding a new document tailored to C# (.NET) projects and onboarding to Codespaces.

### Check off the following:
- [ ] I have reviewed my changes in staging. (look for the **deploy-to-heroku** link in your pull request, then click **View deployment**)
- [ ] For content changes, I have reviewed the [localization checklist](https://github.com/github/docs/blob/main/contributing/localization-checklist.md)
- [ ] For content changes, I have reviewed the [Content style guide for GitHub Docs](https://github.com/github/docs/blob/main/contributing/content-style-guide.md).

### Writer impact (This section is for GitHub staff members only):

- [ ] This pull request impacts the contribution experience
  - [ ] I have added the 'writer impact' label
  - [ ] I have added a description and/or a video demo of the changes below (eg. a "before and after video")

### CODE REVIEW

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+---
+
+
+### Introduction
+
+This guide shows you how to set up your C# (.NET) project in {% data variables.product.prodname_codespaces %}. It will take you through an example of opening your project in a codespace, and adding and modifying a dev container configuration from a template.
+
+#### Prerequisites 
+
+- You should have an existing C# (.NET) project in a repository on {% data variables.product.prodname_dotcom_the_website %}. If you don't have a project, you can try this tutorial with the following example: https://github.com/2percentsilk/dotnet-quickstart.
```
Comment (2percentsilk) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:19]: Open to other sample repos - we want it to be relatively simple so I just used `dotnet new` to generate this application.

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+1. In the Explorer, expand the `.devcontainer` folder and select the `devcontainer.json` file from the tree to open it. 
+
+  !["Codespaces: Rebuild Container" in the command palette](/assets/images/help/codespaces/devcontainers-options.png)  
+
+2. Update your the `extensions` list in your `devcontainer.json` file to add a few extensions that are useful when working with your project.
+   
+  ```json{:copy} 
+  "extensions": [
+		  "ms-dotnettools.csharp",
```
Comment (2percentsilk) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:181]: Any extensions you might suggest that are common/helpful?

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+### Step 1: Open your project in a codespace
+
+1. Navigate to your project's repository. Use the {% octicon "download" aria-label="The download icon" %} **Code** drop-down menu, and select **Open with Codespaces**. If you don’t see this option, your project isn’t available for {% data variables.product.prodname_codespaces %}.
+  
+  ![Open with Codespaces button](/assets/images/help/codespaces/open-with-codespaces-button.png) 
+
+2. To create a new codespace, click {% octicon "plus" aria-label="The plus icon" %} **New codespace**.
+  ![New codespace button](/assets/images/help/codespaces/new-codespace-button.png)
+
+When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including Node.js, JavaScript, Typescript, nvm, npm, and yarn. It also includes a common set of tools like git, wget, rsync, openssh, and nano.
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:31]:
```suggestion
When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including .NET. It also includes a common set of tools like git, wget, rsync, openssh, and nano. 
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+1. Navigate to your project's repository. Use the {% octicon "download" aria-label="The download icon" %} **Code** drop-down menu, and select **Open with Codespaces**. If you don’t see this option, your project isn’t available for {% data variables.product.prodname_codespaces %}.
+  
+  ![Open with Codespaces button](/assets/images/help/codespaces/open-with-codespaces-button.png) 
+
+2. To create a new codespace, click {% octicon "plus" aria-label="The plus icon" %} **New codespace**.
+  ![New codespace button](/assets/images/help/codespaces/new-codespace-button.png)
+
+When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including Node.js, JavaScript, Typescript, nvm, npm, and yarn. It also includes a common set of tools like git, wget, rsync, openssh, and nano. 
+
+You can customize your codespace by adjusting the amount of vCPUs and RAM], [adding dotfiles to personalize your environment](/codespaces/setting-up-your-codespace/personalizing-codespaces-for-your-account), or by modifying the tools and scripts installed.
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:33]:
```suggestion
You can customize your codespace by adjusting the amount of vCPUs and RAM, [adding dotfiles to personalize your environment](/codespaces/setting-up-your-codespace/personalizing-codespaces-for-your-account), or by modifying the tools and scripts installed. 
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+The next section shows you how to modify your tools by adding a dev container.
+
+### Step 2: Add a dev container to your codespace from a template  
+
+The default codespaces container comes with the latest .NET version and common tools preinstalled. However, we encourage you to set up a custom container so you can tailor the tools and scripts that run as part of codespace creation to your project's needs and ensure a fully reproducible environment for all {% data variables.product.prodname_codespaces %} users in your repository.
+
+To set up your project with a custom container, you will need to use a `devcontainer.json` file to define the environment. In {% data variables.product.prodname_codespaces %} you can add this either from a template or you can create your own. For more information on dev containers, see [Configuring your codespace](/codespaces/setting-up-your-codespace/configuring-codespaces-for-your-project).
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:43]:
```suggestion
To set up your project with a custom container, you will need to use a `devcontainer.json` file to define the environment. In {% data variables.product.prodname_codespaces %} you can add this either from a template or you can create your own. For more information on dev containers, see "[Configuring Codespaces for your project
](/codespaces/setting-up-your-codespace/configuring-codespaces-for-your-project)."
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+  ![New codespace button](/assets/images/help/codespaces/new-codespace-button.png)
+
+When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including Node.js, JavaScript, Typescript, nvm, npm, and yarn. It also includes a common set of tools like git, wget, rsync, openssh, and nano. 
+
+You can customize your codespace by adjusting the amount of vCPUs and RAM], [adding dotfiles to personalize your environment](/codespaces/setting-up-your-codespace/personalizing-codespaces-for-your-account), or by modifying the tools and scripts installed. 
+
+{% data variables.product.prodname_codespaces %} uses a file called `devcontainer.json` to store configurations. On launch {% data variables.product.prodname_codespaces %} uses the file to install any tools, dependencies, or other set up that might be needed for the project. For more information, see "[Configuring Codespaces for your project](/codespaces/setting-up-your-codespace/configuring-codespaces-for-your-project)."
+
+The next section shows you how to modify your tools by adding a dev container.
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:37]:
```suggestion
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+### Step 2: Add a dev container to your codespace from a template  
+
+The default codespaces container comes with the latest .NET version and common tools preinstalled. However, we encourage you to set up a custom container so you can tailor the tools and scripts that run as part of codespace creation to your project's needs and ensure a fully reproducible environment for all {% data variables.product.prodname_codespaces %} users in your repository.
+
+To set up your project with a custom container, you will need to use a `devcontainer.json` file to define the environment. In {% data variables.product.prodname_codespaces %} you can add this either from a template or you can create your own. For more information on dev containers, see [Configuring your codespace](/codespaces/setting-up-your-codespace/configuring-codespaces-for-your-project). 
+
+This example guides you through adding a `devcontainer.json` file from a template.
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:45]:
```suggestion
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+  - **Args**
+    - **Variant**: We only have one build argument here which is the node variant we want to use which is passed into our Dockerfile.
+- **Settings** - These are {% data variables.product.prodname_vscode %} settings we wish to set
+  - **Terminal.integrated.shell.linux** - While bash is the default here, we could use zsh for example by modifying this.
+- **Extensions** - These are extensions included by default.
+  - **ms-python.python** - The Microsoft Python extension provides rich support for the Python language (for all actively supported versions of the language: >=3.6), including features such as IntelliSense, linting, debugging, code navigation, code formatting, refactoring, variable explorer, test explorer, and more.
+- **forwardPorts** - By default we can forward a port, like port 3000, but these will also forward automatically
+- **postCreateCommand** - If we want to run anything after we land in our codespace that’s not defined in our Dockerfile, like `dotnet restore`, we can do that here.
+- **remoteUser** - We’re running as the vscode user, but you can optionally set this to root
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:139]:
```suggestion
- **Name** - You can name our dev container anything, this is just the default.
- **Build** - The build properties.
  - **Dockerfile** - In the build object, `dockerfile` is a reference to the Dockerfile that was also added from the template.
  - **Args**
    - **Variant**: This file only contains one build argument, which is the .NET Core version that we want to use.
- **Settings** - These are {% data variables.product.prodname_vscode %} settings.
  - **Terminal.integrated.shell.linux** - While bash is the default here, you could use other terminal shells by modifying this.
- **Extensions** - These are extensions included by default.
  - **ms-dotnettools.csharp** - The Microsoft C# extension provides rich support for developing in C#, including features such as IntelliSense, linting, debugging, code navigation, code formatting, refactoring, variable explorer, test explorer, and more.
- **forwardPorts** - Any ports listed here will be forwarded automatically.
- **postCreateCommand** - If you want to run anything after you land in your codespace that’s not defined in the Dockerfile, like `dotnet restore`, you can do that here.
- **remoteUser** - By default, you’re running as the vscode user, but you can optionally set this to root.
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+  "postCreateCommand": "dotnet restore",
+  ```
+
+4. To rebuild your container and apply the devcontainer.json changes, access the command palette (`shift command P` / `shift control P`), then start typing "rebuild". Click **Codespaces: Rebuild Container**. 
+
+  ![Rebuild container option](/assets/images/help/codespaces/codespaces-rebuild.png)
+
+  Rebuilding inside your codespace ensures your changes work as expected before you commit the changes to the repository. If something does result in a failure, you’ll be placed in a codespace with a recovery container that you can rebuild from to keep adjusting your container.
+
+5. Check your changes were successfully applied by verifying the Code Spell Checker extension was installed.
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:199]:
```suggestion
5. Check your changes were successfully applied by verifying the "Code Spell Checker" extension was installed.
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+   
+![Extensions list](/assets/images/help/codespaces/dotnet-extensions.png)
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:201]:
```suggestion
    ![Extensions list](/assets/images/help/codespaces/dotnet-extensions.png) 
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+  ![Explorer with commit message added](/assets/images/help/codespaces/codespaces-commit-commit-message.png)  
+
+4. To create a PR, click the PR icon. 
+
+  ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-pr-button.png)  
+
+5. Select the branches that you want to merge into, then click **Create**.
+
+  ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-pr.png)
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:235]:
```suggestion
{% data reusables.codespaces.committing-link-to-procedure %} 
```

File: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
+---
+title: Getting started with your C# (.NET) project in Codespaces
+shortTitle: Getting started with your C# (.NET) project
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md:3]:
```suggestion
title: Getting started with your C# (.NET) project in Codespaces
shortTitle: Getting started with your C# (.NET) project
allowTitleToDifferFromFilename: true
```

File: content/codespaces/getting-started-with-codespaces/index.md [hunk 13]
```diff
@@ -7,5 +7,6 @@ versions:
 
 {% link_with_intro /getting-started-with-your-nodejs-project-in-codespaces %}
 
-{% link_with_intro /getting-started-with-your-python-project-in-codespaces %}
+{% link_with_intro /getting-started-with-your-dotnet-project-in-codespaces %}
```
Comment (timeyoutakeit) [inline content/codespaces/getting-started-with-codespaces/index.md:10]:
```suggestion
{% link_with_intro /getting-started-with-your-dotnet-project %}
```

**Review by timeyoutakeit — COMMENTED** [review]
This is looking good 🎉 

There was a couple of issues with the filename and title being different, but hopefully the suggestion should fix it.

**Review by timeyoutakeit — APPROVED** [review]

**Pull request merged**